File: app/src/main/java/com/tunjid/fingergestures/viewholders/MapperViewHolder.java
package com.tunjid.fingergestures.viewholders;

import android.content.Context;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.tunjid.androidbootstrap.core.text.SpanBuilder;
import com.tunjid.fingergestures.R;
import com.tunjid.fingergestures.adapters.AppAdapter;
import com.tunjid.fingergestures.billing.PurchasesManager;
import com.tunjid.fingergestures.fragments.ActionFragment;
import com.tunjid.fingergestures.gestureconsumers.GestureMapper;

import static com.tunjid.fingergestures.App.accessibilityServiceEnabled;
import static com.tunjid.fingergestures.App.canWriteToSettings;
import static com.tunjid.fingergestures.activities.MainActivity.ACCESSIBILITY_CODE;
import static com.tunjid.fingergestures.activities.MainActivity.SETTINGS_CODE;
import static com.tunjid.fingergestures.gestureconsumers.GestureMapper.DOWN_GESTURE;
import static com.tunjid.fingergestures.gestureconsumers.GestureMapper.GestureDirection;
import static com.tunjid.fingergestures.gestureconsumers.GestureMapper.LEFT_GESTURE;
import static com.tunjid.fingergestures.gestureconsumers.GestureMapper.RIGHT_GESTURE;
import static com.tunjid.fingergestures.gestureconsumers.GestureMapper.UP_GESTURE;

public class MapperViewHolder extends AppViewHolder {

    private final TextView title;
    private final TextView subtitle;

    private final GestureMapper mapper;
    @GestureDirection private final String direction;
    @GestureDirection private final String doubleDirection;

    public MapperViewHolder(View itemView, @GestureDirection String direction,
                            AppAdapter.AppAdapterListener listener) {
        super(itemView, listener);
        mapper = GestureMapper.getInstance();

        this.direction = direction;
        this.doubleDirection = mapper.doubleDirection(direction);

        title = itemView.findViewById(R.id.title);
        subtitle = itemView.findViewById(R.id.sub_title);

        title.setOnClickListener(view -> onClick(direction));

        setIcon(itemView.findViewById(R.id.icon), direction);
    }

    @Override
    public void bind() {
        super.bind();
        if (!accessibilityServiceEnabled()) adapterListener.requestPermission(ACCESSIBILITY_CODE);
        if (!canWriteToSettings()) adapterListener.requestPermission(SETTINGS_CODE);

        title.setText(getFormattedText(direction, mapper.getMappedAction(direction)));
        subtitle.setText(getFormattedText(doubleDirection, mapper.getMappedAction(doubleDirection)));

        subtitle.setOnClickListener(view -> {
            boolean notPremium = PurchasesManager.getInstance().isNotPremium();
            if (notPremium) goPremium(R.string.premium_prompt_double_swipe);
            else onClick(doubleDirection);
        });
    }

    private void onClick(@GestureDirection String direction) {
        adapterListener.showBottomSheetFragment(ActionFragment.directionInstance(direction));
    }

    private CharSequence getFormattedText(@GestureDirection String direction, String text) {
        GestureMapper mapper = GestureMapper.getInstance();
        Context context = itemView.getContext();
        return SpanBuilder.format(context.getString(R.string.mapper_format),
                new SpanBuilder(context, mapper.getDirectionName(direction)).bold().build(),
                text);
    }

    private void setIcon(ImageView icon, @GestureDirection String gesture) {
        switch (gesture) {
            case UP_GESTURE:
                icon.setImageResource(R.drawable.ic_keyboard_arrow_up_white_24dp);
                break;
            case DOWN_GESTURE:
                icon.setImageResource(R.drawable.ic_keyboard_arrow_down_white_24dp);
                break;
            case LEFT_GESTURE:
                icon.setImageResource(R.drawable.ic_chevron_left_white_24dp);
                break;
            case RIGHT_GESTURE:
                icon.setImageResource(R.drawable.ic_chevron_right_white_24dp);
                break;
        }
    }
}

File: app/src/main/java/com/tunjid/fingergestures/viewholders/MapperViewHolder.kt
/*
 * Copyright (c) 2017, 2018, 2019 Adetunji Dahunsi.
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.tunjid.fingergestures.viewholders

import android.view.ViewGroup
import com.tunjid.androidx.core.text.bold
import com.tunjid.androidx.core.text.formatSpanned
import com.tunjid.androidx.recyclerview.viewbinding.BindingViewHolder
import com.tunjid.androidx.recyclerview.viewbinding.viewHolderDelegate
import com.tunjid.androidx.recyclerview.viewbinding.viewHolderFrom
import com.tunjid.fingergestures.R
import com.tunjid.fingergestures.accessibilityServiceEnabled
import com.tunjid.fingergestures.ui.main.Item
import com.tunjid.fingergestures.canWriteToSettings
import com.tunjid.fingergestures.databinding.ViewholderMapperBinding
import com.tunjid.fingergestures.ui.popup.ActionFragment
import com.tunjid.fingergestures.gestureconsumers.GestureDirection
import com.tunjid.fingergestures.models.Input

private var BindingViewHolder<ViewholderMapperBinding>.item by viewHolderDelegate<Item.Mapper>()

fun ViewGroup.mapper() = viewHolderFrom(ViewholderMapperBinding::inflate).apply {
    binding.title.setOnClickListener { item.input.accept(Input.UiInteraction.ShowSheet(ActionFragment.directionInstance(item.direction))) }
    binding.subTitle.setOnClickListener {
        item.input.accept(
            if (item.canUseDoubleSwipes) Input.UiInteraction.ShowSheet(ActionFragment.directionInstance(item.doubleDirection))
            else Input.UiInteraction.GoPremium(R.string.premium_prompt_double_swipe)
        )
    }
}

fun BindingViewHolder<ViewholderMapperBinding>.bind(item: Item.Mapper) = binding.run {
    this@bind.item = item

    icon.setImageResource(when (item.direction) {
        GestureDirection.Up,
        GestureDirection.DoubleUp -> R.drawable.ic_keyboard_arrow_up_white_24dp
        GestureDirection.Down,
        GestureDirection.DoubleDown -> R.drawable.ic_app_dock_24dp
        GestureDirection.Left,
        GestureDirection.DoubleLeft -> R.drawable.ic_chevron_left_white_24dp
        GestureDirection.Right,
        GestureDirection.DoubleRight -> R.drawable.ic_chevron_right_white_24dp
    })

    if (!itemView.context.accessibilityServiceEnabled) item.input.accept(Input.Permission.Request.Accessibility)
    if (!itemView.context.canWriteToSettings) item.input.accept(Input.Permission.Request.Settings)

    fun getFormattedText(directionName: String, text: String): CharSequence =
        itemView.context.getString(R.string.mapper_format).formatSpanned(
            directionName.bold(),
            text
        )

    title.text = getFormattedText(item.gesturePair.singleGestureName, item.gesturePair.singleActionName)
    subTitle.text = getFormattedText(item.gesturePair.doubleGestureName, item.gesturePair.doubleActionName)
}
